File: lesson-3_dynamic-programming/the-longest-path/recurrence/main.mjs
import { stdin as input, stdout as output } from 'node:process'
import * as readline from 'node:readline/promises'

const rl = readline.createInterface({ input, output })

const n = await rl.question('')
const t = Array.from({ length: n }, () => [])
const theLongest = Array.from({ length: n }, () => [])
for (let i = 0; i < n; i++) {
    const tmp = (await rl.question('')).split(' ')
    for (const [j, v] of tmp.entries()) {
        t[i][j] = Number(v)
    }
}
for (let i = 0; i < n; i++) {
    theLongest[n - 1][i] = t[n - 1][i]
}
for (let i = n - 2; i >= 0; i--) {
    for (const [j, v] of t[i].entries()) {
        theLongest[i][j] = Math.max(theLongest[i + 1][j], theLongest[i + 1][j + 1]) + v
    }
}
console.log(theLongest[0][0])
rl.close()

File: lesson-3_dynamic-programming/the-longest-path/recurrence/main.kt
import kotlin.math.max

fun main() {
    val n = readLine()!!.toInt()
    val t = Array(n) { emptyArray<Int>() }
    val theLongest = Array(n) { emptyArray<Int>() }
    for (i in 0 until n) {
        t[i] = Array(i + 1) { 0 }
        theLongest[i] = Array(i + 1) { 0 }

        val tmp = readLine()!!.split(' ')
        for ((j, v) in tmp.withIndex()) {
            t[i][j] = v.toInt()
        }
    }
    for (i in 0 until n) {
        theLongest[n - 1][i] = t[n - 1][i]
    }
    for (i in n - 2 downTo 0) {
        for ((j, v) in t[i].withIndex()) {
            theLongest[i][j] = max(theLongest[i + 1][j], theLongest[i + 1][j + 1]) + v
        }
    }
    println(theLongest[0][0])
}
